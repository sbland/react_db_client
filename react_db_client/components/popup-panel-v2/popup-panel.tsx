import { Uid } from '@react_db_client/constants.client-types';
import React, { ReactNode } from 'react';
import ReactDOM from 'react-dom';

import {
  EPopupRegisterAction,
  PopupPanelContext,
} from './popup-panel-provider';
import {
  PopupPanelClosePanelStyle,
  PopupPanelContentPanelStyle,
  PopupPanelWrapStyle,
} from './style';

export interface PopupPanelProps extends IPopupPanelProps {
  // Kept for back compatability
}

export interface IPopupPanelProps {
  /**
   * a node to be rendered in the special component.
   */
  id: Uid;
  renderWhenClosed?: boolean;
  children: ReactNode;
  popupRoot?: string | HTMLElement;
  deleteRootOnUnmount?: boolean;
  zIndex?: number;
  onClose?: () => void;
}

export interface IPopupPanelRenderProps {
  id: Uid;
  renderWhenClosed?: boolean;
  children: ReactNode;
  popupRoot?: string | HTMLElement;
  deleteRootOnUnmount?: boolean;
  zIndex?: number;
  baseZIndex: number;
  z: number;
  open?: boolean;
  closePopup: (Uid) => void;
  root: HTMLElement;
}

export const PopupPanelRender = ({
  id,
  baseZIndex,
  z,
  open,
  closePopup,
  renderWhenClosed,
  children,
  root,
}: IPopupPanelRenderProps) => {
  return ReactDOM.createPortal(
    <PopupPanelWrapStyle
      data-testid={`popupPanel_${id}`}
      style={{
        zIndex: baseZIndex + z * 10,
        display: open ? 'inherit' : 'none',
      }}
    >
      <PopupPanelClosePanelStyle
        onKeyDown={(e) => e.key === 'Escape' && closePopup(id)}
        onClick={() => closePopup(id)}
        aria-label="Close the popup"
        data-testid={'popupPanel_closeBtn'}
      />
      <PopupPanelContentPanelStyle
        data-testid={'popupPanel_content'}
        isOpen={open || false}
        onKeyDown={(e) => e.key === 'Escape' && closePopup(id)}
      >
        {(renderWhenClosed || open) && children}
      </PopupPanelContentPanelStyle>
    </PopupPanelWrapStyle>,
    root
  );
};

export function PopupPanel({
  id,
  renderWhenClosed,
  children,
  popupRoot,
  deleteRootOnUnmount,
  zIndex,
  onClose,
}: IPopupPanelProps) {
  const { dispatchPopupRegister, state } = React.useContext(PopupPanelContext);
  const { popupRegister, baseZIndex } = state;

  const [open, setOpen] = React.useState(false);
  const [root, setRoot] = React.useState<HTMLElement | null>(null);
  const [localZ, setLocalZ] = React.useState(-1);

  React.useEffect(() => {
    dispatchPopupRegister({
      type: EPopupRegisterAction.REGISTER_POPUP,
      args: {
        id,
        root: popupRoot,
        deleteRootOnUnmount,
        z: zIndex,
      },
    });
  }, [id, popupRoot, zIndex, dispatchPopupRegister, deleteRootOnUnmount]);

  React.useEffect(() => {
    if (popupRegister[id]?.open && !open) {
      setRoot(popupRegister[id].root || null);
      setLocalZ(popupRegister[id].z || -1);
      setOpen(true);
    }
    if (!popupRegister[id]?.open && open) {
      if (onClose) onClose();
      setOpen(false);
    }
  }, [id, open, popupRegister, dispatchPopupRegister, onClose]);

  React.useEffect(() => {
    return function cleanup() {
      dispatchPopupRegister({
        type: EPopupRegisterAction.DEREGISTER_POPUP,
        args: id,
      });
    };
  }, [dispatchPopupRegister]);

  const handleClose = React.useCallback(() => {
    dispatchPopupRegister({
      type: EPopupRegisterAction.CLOSE_POPUP,
      args: id,
    });
  }, [id, dispatchPopupRegister]);

  if (!root) return <></>;
  return (
    <PopupPanelRender
      id={id}
      baseZIndex={baseZIndex}
      z={localZ}
      open={open}
      closePopup={handleClose}
      renderWhenClosed={renderWhenClosed}
      children={children}
      root={root}
    />
  );
}
