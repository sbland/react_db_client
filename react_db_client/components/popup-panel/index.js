export { PopupPanel, PopupPanelConnector } from './popup-panel';
