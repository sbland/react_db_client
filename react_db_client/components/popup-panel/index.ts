export { PopupPanel, PopupPanelConnector } from './popup-panel';
export type { IPopupPanelProps } from './popup-panel';
