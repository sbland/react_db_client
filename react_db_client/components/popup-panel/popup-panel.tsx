import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';
import { useGetRoot } from '@react_db_client/helpers.html-helpers';
import {
  PopupPanelCloseBtn,
  PopupPanelContentStyle,
  PopupPanelOverlay,
  PopupPanelStyle,
  PopupPanelTitle,
  PopupPanelTopBar,
  PopupPanelWrapStyle,
} from './style';

export interface IPopupPanelProps {
  id: string;
  title: string;
  isOpen: boolean;
  handleClose: () => void;
  className?: string;
  renderWhenClosed?: boolean;
  children: React.ReactNode;
  popupRoot?: any;
}

let popupCount = 1;
/**
 * Popup panel
 * Make sure to add `<div id="popup-root"></div>` to `index.html`
 * Uses ReactDom.createPortal to place the component in different DOM root
 *
 * @param {*} {
 *   children,
 * }
 * @returns
 */
export const PopupPanel = ({
  id,
  title,
  isOpen,
  handleClose,
  className,
  renderWhenClosed,
  children,
  popupRoot,
}: IPopupPanelProps) => {
  const [z] = useState(popupCount);
  const _popupRoot = useGetRoot(popupRoot || id);
  // const _popupRoot = getRoot(popupRoot || id);

  useEffect(() => {
    popupCount += 1;
    return () => {
      popupCount -= 1;
    };
  }, []);
  if (!isOpen) return <></>;

  const classNames = [className, `popupid_${id}`].filter((f) => f).join(' ');

  if (!_popupRoot) return <div className={classNames}>Missing Popup Root</div>;
  return ReactDOM.createPortal(
    // eslint-disable-next-line jsx-a11y/no-static-element-interactions
    <PopupPanelWrapStyle
      className={`popupPanelWrap ${classNames}`}
      style={{
        zIndex: z * 10,
      }}
      data-testid={`popupid_${id}`}
      onKeyDown={(e) => e.key === 'Escape' && handleClose()}
      // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
      // tabIndex="0"
    >
      <PopupPanelStyle className="popupPanel" data-testid="rdc-popupPanel">
        <PopupPanelContentStyle className="popupPanel_content">
          {(renderWhenClosed || isOpen) && children}
        </PopupPanelContentStyle>
        <PopupPanelTopBar className="popupPanel_topBar">
          <PopupPanelTitle className="popupPanel_title">{title}</PopupPanelTitle>
          <PopupPanelCloseBtn className="popupPanel_closeBtn" type="button" onClick={handleClose}>
            X
          </PopupPanelCloseBtn>
        </PopupPanelTopBar>
      </PopupPanelStyle>
      {/* eslint-disable-next-line max-len */}
      {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */}
      <PopupPanelOverlay className="popupPanel_overlay" onClick={handleClose} />
    </PopupPanelWrapStyle>,
    _popupRoot
  );
};

PopupPanel.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.node,
  children: PropTypes.any.isRequired,
  isOpen: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  renderWhenClosed: PropTypes.bool,
  popupRoot: PropTypes.any,
};

PopupPanel.defaultProps = {
  renderWhenClosed: false,
  popupRoot: null,
};

// Uses React HOC pattern
export const PopupPanelConnector = (
  Component,
  root,
  alwaysOpen,
  closeProp = 'handleClose',
  propsOverrides = {}
) => {
  return (props) => {
    const propsMerged = { ...props, ...propsOverrides };
    const { className, isOpen, title, id } = propsMerged;
    const handleClose = props[closeProp];
    return (
      <PopupPanel
        handleClose={handleClose}
        isOpen={alwaysOpen || isOpen}
        title={title}
        popupRoot={root}
        className={className}
        id={id || root}
      >
        <Component {...props} />
      </PopupPanel>
    );
  };
};
