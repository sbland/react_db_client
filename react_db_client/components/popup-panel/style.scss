$unit: var(--unit, 1rem);
$white-off: #dddddd;
$primaryColour: var(--primary-color, #888);
$grey-40: #bbbbbb;


.popupPanelWrap {
  position: fixed;

  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  overflow-y: auto;
}

.popupPanel {
  position: fixed;
  left: ($unit * 1);
  right: ($unit * 1);
  top: ($unit * 1);
  bottom: ($unit * 1);
  background: $white-off;
  z-index: 10;

  padding: ($unit * 2);
  border: 1px $primaryColour solid;
  box-shadow: 1px 1px 10px 1px rgba(100,100,100,0.3);
  height: 95%;
  // overflow-y: auto;
}

.popupPanel_overlay {
  position: fixed;
  // background: black;
  opacity: 0.4;
  cursor: pointer;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  background:url(
    data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAcAAAAHCAYAAADEUlfTAAAAMElEQVQImXXJsREAIAwDMS/kZX47b0sFByQpVElAOrajKYCaO0re8eQfJ7sAoimALAAwSVfRiD/CAAAAAElFTkSuQmCC
  ) repeat;
}

.popupPanel_closeBtn {
  position: absolute;
  top: -10px;
  right: -10px;
  top: -10px;
  border-radius: 10px;
  border: none;
}

.popupPanel_content {
  margin-bottom: $unit;
  position: absolute;
  overflow-y: auto;
  top: ($unit * 1);
  right:0;
  left: 0;
  bottom: 0;
  padding: ($unit);
}

.popupPanel_topBar {
  position: absolute;
  top: 0;
  right:0;
  left: 0;
  height: ($unit * 2);
  background: $white-off;
}

.popupPanel_title {
  margin-left: $unit;
  color: $grey-40;
}
