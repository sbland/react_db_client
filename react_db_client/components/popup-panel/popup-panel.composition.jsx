import React, { useState } from 'react';
import { PopupPanel, PopupPanelConnector } from './popup-panel';

export const BasicPopupPanel = () => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <div className="">
      <div className="">
        <button
          type="button"
          className={isOpen ? 'button-one' : 'button-two'}
          onClick={() => setIsOpen(true)}
        >
          Open
        </button>
        <PopupPanel
          id="demo-id"
          isOpen={isOpen}
          handleClose={() => {
            setIsOpen(false);
          }}
          popupRoot="root"
        >
          <div className="" style={{ background: 'red' }}>
            Hello
          </div>
        </PopupPanel>
      </div>
    </div>
  );
};

const DemoInnerComponent = (props) => {
  return <div>Hello</div>;
};
const WrappedComponent = PopupPanelConnector(DemoInnerComponent, 'root', true, 'onCancel');

export const PopupPanelConnectorDemo = () => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <div className="">
      <div className="">
        <button
          type="button"
          className={isOpen ? 'button-one' : 'button-two'}
          onClick={() => setIsOpen(true)}
        >
          Open
        </button>
        {isOpen && <WrappedComponent onCancel={() => setIsOpen(false)} />}
      </div>
    </div>
  );
};
