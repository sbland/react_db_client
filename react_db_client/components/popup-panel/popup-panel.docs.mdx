---
description: 'A PopupPanel component.'
labels: ['ui', 'popup']
---

import { PopupPanel } from './popup-panel';

## React Component for creating a popup panel

Wraps children components in a floating window.