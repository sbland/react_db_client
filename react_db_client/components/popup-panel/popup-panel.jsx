import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';

import './style.scss';

const getRoot = (inputRoot) => {
  let root = null;
  if (typeof inputRoot == 'object') root = inputRoot;
  if (typeof inputRoot == 'string') root = document.getElementById(inputRoot);
  if (!root) {
    root = document.createElement('div');
    root.setAttribute('id', inputRoot || '_root');
    document.body.appendChild(root);
  }
  return root;
};

let popupCount = 1;
/**
 * Popup panel
 * Make sure to add `<div id="popup-root"></div>` to `index.html`
 * Uses ReactDom.createPortal to place the component in different DOM root
 *
 * @param {*} {
 *   children,
 * }
 * @returns
 */
export const PopupPanel = ({
  id,
  title,
  isOpen,
  handleClose,
  className,
  renderWhenClosed,
  children,
  popupRoot,
}) => {
  const [z] = useState(popupCount);
  const _popupRoot = getRoot(popupRoot || id);

  useEffect(() => {
    popupCount += 1;
    return () => {
      popupCount -= 1;
    };
  }, []);
  if (!isOpen) return <></>;

  const classNames = [className, `popupid_${id}`].filter((f) => f).join(' ');

  if (!_popupRoot) return <div className={classNames}>Missing Popup Root</div>;
  return ReactDOM.createPortal(
    // eslint-disable-next-line jsx-a11y/no-static-element-interactions
    <div
      className={`popupPanelWrap ${classNames}`}
      style={{
        zIndex: z * 10,
      }}
      onKeyDown={(e) => e.key === 'Escape' && handleClose()}
      // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
      tabIndex="0"
    >
      <div className="popupPanel">
        <div className="popupPanel_content">{(renderWhenClosed || isOpen) && children}</div>
        <div className="popupPanel_topBar">
          <div className="popupPanel_title">{title}</div>
          <button className="popupPanel_closeBtn" type="button" onClick={handleClose}>
            X
          </button>
        </div>
      </div>
      {/* eslint-disable-next-line max-len */}
      {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */}
      <div className="popupPanel_overlay" onClick={handleClose} />
    </div>,
    _popupRoot
  );
};

PopupPanel.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.node,
  children: PropTypes.any.isRequired,
  isOpen: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  renderWhenClosed: PropTypes.bool,
  popupRoot: PropTypes.any,
};

PopupPanel.defaultProps = {
  renderWhenClosed: false,
  popupRoot: null,
};

// Uses React HOC pattern
export const PopupPanelConnector = (
  Component,
  root,
  alwaysOpen,
  closeProp = 'handleClose',
  propsOverrides = {}
) => {
  return (props) => {
    const propsMerged = { ...props, ...propsOverrides };
    const { className, isOpen, title, id } = propsMerged;
    const handleClose = props[closeProp];
    return (
      <PopupPanel
        handleClose={handleClose}
        isOpen={alwaysOpen || isOpen}
        title={title}
        popupRoot={root}
        className={className}
        id={id || root}
      >
        <Component {...props} />
      </PopupPanel>
    );
  };
};
