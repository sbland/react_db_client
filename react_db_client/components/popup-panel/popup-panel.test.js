import '@samnbuk/react_db_client.testing.enzyme-setup';
import React from 'react';
import { shallow, mount } from 'enzyme';
import * as compositions from './popup-panel.composition';

const defaultProps = {
  //
};

describe('popup panel', () => {
  describe('Compositions', () => {
    Object.entries(compositions).forEach(([name, Composition]) => {
      test(name, () => {
        const component = mount(<Composition />);
        expect(component.debug()).toMatchSnapshot();
      });
    });
  });
});
