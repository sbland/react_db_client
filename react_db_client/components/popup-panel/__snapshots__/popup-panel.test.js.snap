// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`popup panel Compositions BasicPopupPanel 1`] = `
"<BasicPopupPanel>
  <div className=\\"\\">
    <div className=\\"\\">
      <button type=\\"button\\" className=\\"button-two\\" onClick={[Function: onClick]}>
        Open
      </button>
      <PopupPanel id=\\"demo-id\\" isOpen={false} handleClose={[Function: handleClose]} popupRoot=\\"root\\" renderWhenClosed={false} />
    </div>
  </div>
</BasicPopupPanel>"
`;

exports[`popup panel Compositions PopupPanelConnectorDemo 1`] = `
"<PopupPanelConnectorDemo>
  <div className=\\"\\">
    <div className=\\"\\">
      <button type=\\"button\\" className=\\"button-two\\" onClick={[Function: onClick]}>
        Open
      </button>
    </div>
  </div>
</PopupPanelConnectorDemo>"
`;
